src/lib/FreeC/Backend/Coq/Converter/TypeDecl.hs

```diff
@@ -374,3 +385,644 @@ convertDataDecl (IR.DataDecl _ (IR.DeclIdent _ name) typeVarDecls conDecls) = do
+--                -> Free Shape Pos (List Identity.Shape Identity.Pos
+--                                    (Forest Identity.Shape Identity.Pos b))@
+--
+--   To generate these local function, for every type expression @aij@ in the
+--   constructors of @T@, we collect all types that contain the original type
+--   @T@.
+--   More specifically, a type expression @T' t1 ... tm@ is collected if
+--   @ti = T t1' ... tn'@ for some type expressions @t1', ..., tn'@, or if @ti@
+--   is collected for some @i@.
+--   During this process, any type expression that does not contain @T@ is
+--   replaced by a placeholder variable "_".
+--
+--   We keep track of which types correspond to which function with a map.
+--
+--   The generated functions @fT1, ..., fTn@ for @n@ mutually recursive types
+--   @T1, ... Tn@ are a set of @n@ @Fixpoint@ definitions linked with @with@.
+--   Indirectly recursive types and local functions based on them are computed
+--   for each type.
+--   In this case, a type @T'@ is considered indirectly recursive if it
+--   contains any of the types @T1, ..., Tn@.
+--   Arguments of type @Ti@ can be treated like directly recursive arguments.
+generateTypeclassInstances :: [IR.TypeDecl] -> Converter [Coq.Sentence]
+generateTypeclassInstances dataDecls = do
+  -- The types of the data declaration's constructors' arguments.
+  let argTypes = map (concatMap IR.conDeclFields . IR.dataDeclCons) dataDecls
+  -- The same types where all type synonyms are expanded.
+  argTypesExpanded
+    <- mapM (mapM expandAllTypeSynonyms) argTypes -- :: [[IR.Type]]
+  -- A list where all fully-applied type constructors that do not contain one of the types
+  -- for which we are defining instances and all type variables are replaced with
+  -- the same type variable (an underscore). The list is reversed so its entries are
+  -- in topological order.
+  let reducedTypes = map (nub . reverse . concatMap collectSubTypes)
+        argTypesExpanded
+  -- Like reducedTypes, but with all occurrences of the types for which we are defining
+  -- instances and all type variables removed from the list.
+  -- This leaves exactly the types with indirect recursion, with all non-recursive
+  -- components replaced by underscores.
+  let recTypeList = map
+        (filter (\t -> not (t `elem` declTypes || IR.isTypeVar t))) reducedTypes
+  -- Construct Normalform instances.
+  nfInstances <- buildInstances recTypeList normalformFuncName
+    normalformClassName nfBindersAndReturnType buildNormalformValue
+  -- Construct ShareableArgs instances.
+  shareableArgsInstances <- buildInstances recTypeList shareableArgsFuncName
+    shareableArgsClassName shareArgsBindersAndReturnType buildShareArgsValue
+  return (nfInstances ++ shareableArgsInstances)
+ where
+  -- The (mutually recursive) data types for which we are defining
+  -- instances, converted to types.
+  declTypes :: [IR.Type]
+  declTypes = map dataDeclToType dataDecls
```

Comment:
Since `dataDeclToType` is used nowhere else, you could inline it here, for example, by using list comprehensions. That type variables are replaced by `_` may also be worth mentioning in the comment.

```suggestion
-- The (mutually recursive) data types for which we are defining
-- instances, converted to types.
declTypes :: [IR.Type]
declTypes = [IR.typeConApp NoSrcSpan (IR.typeDeclQName dataDecl)
              (replicate (length (IR.typeDeclArgs dataDecl)) placeholderVar)
            | dataDecl <- dataDecls
            ]
```